README
This is a highly-modified for of the RabbIT proxy from 
http://www.khelekore.org/rabbit.  It has been modified to be more easily
callable and used as a library, as well as removing a lot of unused
(by me, anyway) stuff.


ChangeLog
version 5.0 (20xx-yy-zz) Only pass a proxy-authentication token if it is
    actually set to something. Made it possible to configure
    tcp_no_delay on or off. Do not close connections when a filter
    sends back an unauthorized header.  Move rnio back into rabbit.rnio.
    Simplify the code so it is more easily used as a library, rather
    than a stand-alone application (i.e. around configs, etc).

version 4.11 (2012-02-16) Cleanup of the request logging. Make sure that
    logging is done off the selector threads so that we can call
    external loggers. Added configurable traffic usage loggers so that
    it is easy to log user network traffic usage. Extracted cache
    configuration to its own interface and moved the ncache
    implementation to its own package. The CacheEntry interface was
    cleaned up so that it is easier to implement and easier to use,
    some methods that were only for internal usage were removed. More
    fixes for http headers reading, when the header has been split
    into many different blocks on the network.

version 4.10 (2011-06-29) Fixed a problem with StackOverflow for chunked
    resources with many small chunks. Updated rnio to version
    1.2. Fixed a problem with enlarging an already large data buffer
    data buffer. Fixed a problem that caused some bytes to get lost
    when a data buffer were enlarged. Updated the cache_only config so
    that rabbit starts without warnings. Try to return the data buffer
    in a few more cases. Updated javadocs.

version 4.9 (2011-01-18) Added a jndi context for database resources. Made
    it easier to use external classes for databases by adding custom
    class loading. Stopped using TCP_NO_DELAY. Fixed a problem that
    caused RequestLineTooLongException to be thrown in some rare
    cases. Much updates of the javadoc. Added a filter that blocks
    sites using a sql database table of hosts. Update the rnio
    library.

version 4.8 (2010-10-27) Big javadoc update. Add support for filtering
    of connect requests. Fixed the problem where the log file got
    filled with IllegalStateException when repack was on. Handle
    character encodings and http header content better. Handle semi
    transparent images better when using the java based image
    converter. Updated dnsjava to 2.1.0. Shorten the log from
    "Connection reset by peer". Added addresses to SSLHandlers logs.

version 4.7 (2010-07-25) Fixed the problem 401 and 407 responses when
    using FileTemplateHttpGeneratorFactory. Added a proxy chain
    selector to make it it possible to use different connections
    strategies depending on the request uri. Made it possible to
    specify a ProxyChainFactory in rabbit.conf. Added a
    ProxyChainFactory that uses direct access to internal networks and
    uses a chained proxy for external access. Made sure that the ant
    based build got valid Main-Class and a Class-Path entries and that
    the Help.txt was included in the jar. Updated ant javadoc task so
    that it uses the same arguments as the Makefile. Fixed many
    trivial issues flagged by idea. A lot of javadoc was added or
    updated. Made sure that http header read buffers are expanded once
    before throwing the 414 request too long response. Fixed a problem
    with reading pipelined requests. Removed the clientHandler from
    the different handlers since they did not and must not use it.

version 4.6 (2010-05-24) Reading a broken cache index now results
    in an empty cache instead of a cache with no files, but a big
    size. Made it possible to configure what HttpGenerator to
    use. ImageHandler now detects and unpacks images that have a
    content encoding set to gzip. Added a http generator that uses a
    directory of template pages for the actual error pages (like 400,
    504, ...). Some minor code cleanup in cache.

version 4.5 (2010-03-23) Created rnio as a separate project out of
    the rabbit.nio package and made rabbit use rnio. Use URLDeocder
    instead of home written method. Use utf-8 for meta pages by
    default. Add a few missing javadoc package descriptions and fixed
    a few typos in the javadoc. Renamed Coder to Base64 since that is
    all it handles. Made it possible for filters to see the client
    request body. Fixed some cross site scripting problems with the
    generated error pages. Changed the cache so that it throws
    CacheException if an operation fails.

version 4.4 (2010-01-04) Made it possible to restrict sql authentication to
    one ip only. Moved the image conversion to separate package and
    merge the different image handlers into just one ImageHandler that
    uses the right image converter internally. Made ImageHandler fall
    back to java based image conversion if the external process
    fails. Merged the different proxy authentication filters into one
    filter class and several back-ends. Moved proxy authentication
    caching up to ProxyAuth so that all authenticators can use it.
    Added some logging to the image handler. Setting repack to true and
    compress to false now works. Added a setting "allow_without_auth"
    to the ProxyAuth so that it possible to access some pages even if
    proxy authentication is generally required. Fixed a problem with
    HTTP/0.9 request.

version 4.3 (2009-11-17) Some code cleanups identified by findbugs.
    Changed the handler setup to use full regexps. Updated the handler
    setup so that it catches more images. Updated the sample web
    server so that it is more secure and sends correct mime type in
    more cases. Updated dnsjava to 2.0.7. Added a image handler that
    uses ImageIO/ImageWriter so it is full java and that means that it
    does not have to fork a convert process. Updated some filter
    documentation. Made it possible to configure SQLProxyAuth to cache
    the results.

version 4.2 (2009-08-20) Made it possible to specify what ip rabbit
    should listen on. Made it possible to specify what ip rabbit

src/main/java/rabbit/util/Base64.java
package rabbit.util;

/** This class encodes/decodes stuff to/from the web.
 */
public class Base64 {

    /** don't construct this
     */
    private Base64() {
        // nah.
    }

    /** the base64 characters 
     */
    private static final int[] pr2six = {
            64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
            64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 62, 64, 64, 64, 63,
            52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 64, 64, 64, 64, 64, 64, 64,  0,  1,  2,  3,  4,  5,  6,
            7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64, 64,
            64, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48,
            49, 50, 51, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
            64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
            64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
            64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
            64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
            64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
    };


    /** The base64 characters.
     */
    private static final char[] uu_base64 = {
            'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P',
            'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f',
            'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v',
            'w', 'x', 'y', 'z', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '+', '/'
    };

    /** uudecode(base64) the given string.
     *  useful for decoding basic Authentication 
     * @param base64string the String to decode.
     * @return the decoded string.
     */
    public static String decode(String base64string) {
        final StringBuilder ret = new StringBuilder(base64string.length() * 3 / 4);

        while ((base64string.length() % 4) != 0) {
            base64string += "=";           // that should be safe.
        }
        int i = 0;
        int c1, c2, c3 = 0, c4 = 0;
        while (i < base64string.length() &&
               pr2six[base64string.charAt(i)] <= 63) {
            c1 = pr2six[base64string.charAt(i)];
            c2 = pr2six[base64string.charAt(i + 1)];
            c3 = pr2six[base64string.charAt(i + 2)];
            c4 = pr2six[base64string.charAt(i + 3)];
            ret.append((char) (c1 << 2 | c2 >> 4));
            ret.append((char) ((c2 << 4 | c3 >> 2) % 256));
            ret.append((char) ((c3 << 6 | c4) % 256));
            i += 4;
        }

        if (c3 > 63) {
            ret.setLength(ret.length() - 2);
        } else if (c4 > 63) {
            ret.setLength(ret.length() - 1);
        }
        return ret.toString();
    }

    /** uuencode(base64) the given String.
     *  useful for encoding basic authentication.
     * @param str the String to encode.
     * @return the encoded string.
     */
    public static String encode(final CharSequence str) {
        final StringBuilder ret = new StringBuilder(str.length() * 4 / 3);
        char ch, ch1, ch2, ch3;
        int i;

        for (i = 0; i + 2 < str.length(); i += 3) {
            ch1 = str.charAt(i);
            ch2 = str.charAt(i + 1);
            ch3 = str.charAt(i + 2);
            ch = uu_base64[((ch1 >> 2) & 0x3f)];
            ret.append(ch);

            ch = uu_base64[(((ch1 << 4) & 0x30) | ((ch2 >> 4) & 0xf))];
            ret.append(ch);

            ch = uu_base64[(((ch2 << 2) & 0x3c) | ((ch3 >> 6) & 0x3))];
            ret.append(ch);

            ch = uu_base64[(ch3 & 0x3f)];
            ret.append(ch);
        }

        // are we done yet?
        if (i == str.length()) {
            return ret.toString();
        }

        // no so handle the trailing characters.
        ch1 = str.charAt(i);
        ch2 = str.length() > i + 1 ? str.charAt(i + 1) : (char) 0;

        ch = uu_base64[((ch1 >> 2) & 0x3f)];
        ret.append(ch);

        ch = uu_base64[(((ch1 << 4) & 0x30) | ((ch2 >> 4) & 0xf))];
        ret.append(ch);

        if (str.length() > i + 1) {
            ch = uu_base64[((ch2 << 2) & 0x3c)];
            ret.append(ch);
        } else {
            ret.append('=');
        }
        ret.append('=');
        return ret.toString();
    }
}

